text/0000-wire-reform.md

```diff
@@ -0,0 +1,247 @@
+* `@lwc/wire-service` becomes Lightning Platform-specific for the most part (`register()` method) since anyone can implement the wire adapter protocol.
+
+_Backwards Compatibility Notes:_
+
+* As today, the descriptor installed on the prototype of the component by the `@wire` decorator was identical to `track` when decorating a field. This means that the component author could change the value of the field, and such change will be tracked. This is no longer the case, and even though the author can still change the value, it will not be reactive, causing no side effects on the UI of the element. This is a non-backwards compatible change, but we believe this is a very low risk for something that was not working as expected.
+
+### Wire Adapter Protocol
+
+The formalization of the wire adapter protocol is important because that enables the interoperability aspect of this feature. The adapter's code should not be aware of the component system, or even the application framework. It only cares about very specific hints to produce a stream of data. The following describes the proposed protocol:
+
+```ts
+interface WireAdapter {
+    update(config: ConfigValue, context?: ContextValue);
+    connect();
+    disconnect();
+}
+interface WireAdapterConstructor {
+    new (callback: DataCallback): WireAdapter;
+    configSchema?: Record<string, WireAdapterSchemaValue>;
+    contextSchema?: Record<string, WireAdapterSchemaValue>;
+}
+type DataCallback = (value: any) => void;
+type ConfigValue = Record<string, any>;
+type ContextValue = Record<string, any>;
+type WireAdapterSchemaValue = 'optional' | 'required';
+```
+
+_Notes:_
+
+* Not all environments will support or need context (e.g.: preloading LDS data), but does supporting it can rely on the static field called `contextSchema` to provide the context value when available.
+* Some environments might choose to implement validation rules for `configSchema` and `contextSchema` to guarantee compliance.
+* we favored the `DataCallback` over a promised based on `update()` calls because the callback can be invoked sync and async, but most important, because update might never be called by the environment.
+
+### Semantic changes for `@wire` decorator IDL
+
+There exist a few restrictions and ambiguities with the IDL for the config object in `@wire` decorator declarations. This section will describe the changed semantics. Most use cases of `@wire` are unaffected. 
+
+* `$token` can only appear as the value of a top level member property, e.g.: `@wire(foo, { x: '$prop1' })` will continue to be valid, while `@wire(foo, { x: { y: '$prop1' } })` will throw a compiler error, while today it doesn't but the value is never transformed, and remains as a string value. This is a non-backwards compatible change, but we believe this is a very low risk for something that was not working as expected.
```

Comment:
@caridy I faced the need of having a $token injected into a non top level property. Is there a technical limitation why this can't be done, or is it because of a potential performance concern with object comparison?
On the same subject, I also faced the need of having the $token being replaced as part of a string, like in:
```
    @track offset = 0
    @track limit = 5
    @wire(useFetch, {
        url: '/users?offset=$offset&limit=$limit',
    ...
```
Or, even better, I would love some of these parameters to be functions like the peudo code bellow:
```
    @track offset = 0
    @track limit = 5
    @wire(useFetch, {
        url: url('/users', {offset: '$offset', limit: '$limit'} ),
    ...
```
It probably means that the whole json object should be recreated on any change, and compared to the previous instance. As there are generally few properties, this should not be a performance problem.